### ISSUE TITLE
Extension API for saving/restoring/modifying layout

### ISSUE THREAD

**Issue by milliams** [jupyterlab/jupyterlab#4575]
I'm working on an extension which will need the ability to programmatically alter the 'layout' of JupyterLab based on a config file. This is which files are open and how they are arranged on the screen.

Since the current serialisation of the `ApplicationShell.ILayout` types are bound up in private methods in `application/layoutrestorer.ts` the only way I can get the the layout in a serialisable form is by grabbing it diretly from the state DB:

```typescript
// state = IStateDB;
state.fetch('layout-restorer:data').then((data) => {
  const layout_string = JSON.stringify(data);
  ...download the layout file to the user as text...
});
```

This does work although I'd much rather be able to do something like:

```typescript
const layout_object = app.shell.saveLayout();
const layout = something.dehydrate_layout(layout_object); // Doesn't exist
const layout_string = JSON.stringify(layout);
...download the layout file to the user as text...
```

It's even trickier trying to apply that layout back to JupyterLab at a later point. I'm currently doing something like:

```typescript
// layout_text = contents of file previously downloaded
const layout = JSON.parse(layout_text);

state.save('layout-restorer:data', layout).then(() => {
  restorer.fetch().then((saved: ApplicationShell.ILayout) => {
    app.shell.restoreLayout(saved);
  });
});
```

which doesn't even seem to work most of the time. It would be good to be able to do something like:

```typescript
// layout_text = contents of file previously downloaded
const layout = JSON.parse(layout_text);
const layout_object = something.rehydrate_layout(layout); // Doesn't exist
app.shell.restoreLayout(layout_object);
```

Basically, currently the layout restorer assumes that the source and sink of the layout definition is the state DB but I would like to be able to refine the API a little so that the dehydration/rehydration is separate from the saving to the DB.
